src/terms/nonlocal.jl

```diff
@@ -178,22 +178,36 @@ end
 """
 Build form factors (Fourier transforms of projectors) for an atom centered at 0.
 """
-function build_form_factors(psp, qs)
-    qnorms = norm.(qs)
-    T = real(eltype(qnorms))
-    # Compute position-independent form factors
-    form_factors = zeros(Complex{T}, length(qs), count_n_proj(psp))
-    count = 1
-    for l in 0:psp.lmax, m in -l:l
-        prefac_lm = im^l .* ylm_real.(l, m, qs)
-        n_proj_l = size(psp.h[l + 1], 1)
+function build_form_factors(psp, G_plus_ks)
+    T = real(typeof(norm(first(G_plus_ks))))
+
+    # Precompute radial parts at unique |G| and store them in a hash map for O(1) lookup.
+    pq = Dict{T,Matrix{T}}()
```

Comment:
One thing we should think about a bit is variable names `pq` is not the most clear.